File: JavaScript/232.js
/**
 * Initialize your data structure here.
 */
var MyQueue = function() {
    this.inArr = [];
    this.outArr = [];
};

/**
 * Push element x to the back of queue. 
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function(x) {
    this.inArr.push(x);
};

/**
 * Removes the element from in front of queue and returns that element.
 * @return {number}
 */
MyQueue.prototype.pop = function() {
    if(!this.outArr.length) {
        while(this.inArr.length) {
            this.outArr.push(this.inArr.pop())
        }
    }
    return this.outArr.pop();
};

/**
 * Get the front element.
 * @return {number}
 */
MyQueue.prototype.peek = function() {
    if(!this.outArr.length) {
        while(this.inArr.length) {
            this.outArr.push(this.inArr.pop())
        }
    }
    return this.outArr[this.outArr.length-1];
};

/**
 * Returns whether the queue is empty.
 * @return {boolean}
 */
MyQueue.prototype.empty = function() {
    if(this.inArr.length || this.outArr.length)
        return false;
    else 
        return true;
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */

File: Cpp/232.cpp
class MyQueue {
public:
    stack<int> A;
    stack<int> B;
    /** Initialize your data structure here. */
    MyQueue() {
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        A.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        check();
        int v = B.top();
        B.pop();
        return v;
    }
    
    /** Get the front element. */
    int peek() {
        check();
        return B.top();
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return A.empty() && B.empty();
    }
    void check() {
        if(B.empty()) {
            while(!A.empty()) {
                B.push(A.top());
                A.pop();
            }
        }
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */